prepare_conditioning_dataset.py

```diff
@@ -3,36 +3,70 @@
 from glob import glob
 
 import torch
+from einops import rearrange
+from kornia.filters import gaussian_blur2d, laplacian
 from torch import nn
 from torch.utils.data import DataLoader, Dataset
-from torchvision.io import read_video
+from torchvision.io import read_video, write_video
 from torchvision.models.optical_flow import Raft_Large_Weights, raft_large
 from tqdm import tqdm
 
 
 class FlowPreprocessor(nn.Module):
-    out_channels: int = 2
+    out_channels: int = 3
 
-    def __init__(self) -> None:
+    def __init__(self, smooth: float = 0.05, emphasis: float = 6.0) -> None:
         super().__init__()
+        self.transform = Raft_Large_Weights.DEFAULT.transforms()
         self.raft = raft_large(weights=Raft_Large_Weights.DEFAULT, progress=True).eval()
+        self.smooth = smooth
+        self.emphasis = emphasis
 
+    @torch.no_grad()
     def forward(self, video: torch.Tensor) -> torch.Tensor:
-        video = video.mul(2).sub(1)
-        return self.raft.forward(video[:-1], video[1:])[-1]
+        # prepare and run RAFT
+        prev, curr = video[:-1], video[1:]
+        prev, curr = self.transform(prev, curr)
+        flow = self.raft.forward(prev, curr)[-1]
+
+        # smooth flow
+        sigma = min(flow.shape[3], flow.shape[2]) * self.smooth
+        flow = gaussian_blur2d(flow, kernel_size=5, sigma=(sigma, sigma))
+
+        # prepend zero flow to match video length
+        flow = torch.cat((torch.zeros_like(flow[[0]]), flow))
+
+        # normalize flow by video size
+        flow[:, 0] /= video.shape[3]
+        flow[:, 1] /= video.shape[2]
+
+        # calculate flow laplacian magnitude (rough edge detection)
+        flow_laplacian = laplacian(flow, kernel_size=5)
+        magnitude = flow_laplacian.abs().sum(1, keepdim=True).pow(1 / self.emphasis).mul(2).sub(1)
+
+        # concatenate to 3-channel conditioning
+        conditioning = torch.cat((flow, magnitude), dim=1)
+
+        # emphasize small values
+        conditioning = torch.sign(conditioning) * conditioning.abs().pow(1 / self.emphasis)
```

Comment:
Interesting. I've done similar things with optical flow before as well but I would assume that since we are training with it, this will happen some what automatically?

Reply:
Presumably it will, but a little extra normalization upfront never hurt anybody